File: algorithms/bfs&dfs/partition_to_k_equal_sum_subsets.cc
// https://leetcode.com/problems/partition-to-k-equal-sum-subsets/

// time complexity:

#include <numeric>
#include <vector>

using namespace std;

class Solution {
public:
    bool dfs(int k, int start, int curr) {
        if (curr == target)
            return k == 2 ? true : dfs(k - 1, 0, 0);

        for (int i = start; i < len; i++) {
            if (!used[i] && curr + _nums[i] <= target) {
                used[i] = true;
                if (dfs(k, i + 1, curr + _nums[i]))
                    return true;
                used[i] = false;
            }
        }
        return false;
    }

    bool canPartitionKSubsets(vector<int> &nums, int k) {
        int sum = accumulate(nums.begin(), nums.end(), 0);
        if (sum % k)
            return false;

        sort(nums.begin(), nums.end(), [](const auto &left, const auto &right) { return left < right; });
        len = nums.size(), target = sum / k;
        _nums = nums, used = vector<bool>(len);
        return dfs(k, 0, 0);
    }

private:
    vector<int> _nums;
    vector<bool> used;
    int len, target;
};

File: algorithms/bfs&dfs/partition_to_k_equal_sum_subsets.py
# https://leetcode.com/problems/partition-to-k-equal-sum-subsets/

# time complexity:

class Solution:
    def canPartitionKSubsets(self, nums: list[int], k: int) -> bool:
        if (_sum := sum(nums)) % k:
            return False
        
        nums.sort(reverse=True)
        used, target = [False] * (_len := len(nums)), _sum // k

        def dfs(k, start, curr):
            if curr == target:
                return True if k == 2 else dfs(k - 1, 0, 0)

            for i in range(start, _len):
                if not used[i] and curr + nums[i] <= target:
                    used[i] = True
                    if dfs(k, i + 1, curr + nums[i]):
                        return True
                    used[i] = False
            return False
        
        return dfs(k, 0, 0)

# Though it's slow due to repeating visits, an alternative solution:
# https://leetcode.com/problems/partition-to-k-equal-sum-subsets/discuss/108741/Solution-with-Reference

    def canPartitionKSubsets(self, nums: list[int], k: int) -> bool:
        if (_sum := sum(nums)) % k:
            return False
        
        partition = [_sum // k] * k
        _len = len(nums)
        nums.sort(reverse=True)
        
        def dfs(idx: int) -> bool:
            if idx == _len:
                return True
            for i, p in enumerate(partition):
                if p >= nums[idx]:
                    partition[i] -= nums[idx]
                    if dfs(idx + 1):
                        return True
                    partition[i] += nums[idx]
            return False
        return dfs(0)